### PR TITLE
Factor out implementation header, fix specializations

### LINKED ISSUE #14643: Permit instantiation of complex valued tensors
## Reason
For a milestone involving Fourier transforms of scalar and tensor valued fields it would be very convenient to have instantiations of `RankTwoTensorTempl` on `std::complex<Real>`. 

## Design
Factor out an "implementaion" header to permit custom instantiations in addition to the existing explicit instantiations on `Real` and `DualReal`.

## Impact
Added capability.

### PR DESCRIPTION
**Pull request by dschwen**
Factor out implementation headers to allow for custom type instantiations. Also the specializations were conceptually turned on their heads (where the "general" members were actually _only_ working with `Real` numbers (due to LAPACK calls) and the `DualReal` specializations were actually the general C++ implementations. I fixed that.

Closes #14643

### CODE REVIEW

**Comment by dschwen** [comment]
Note that the failing version was formatted with clang-format-10 (which yields the same results as clang-format-9). Test boxes use clang-format-8 and diffed. Reformated with clang-format-8 now.

**Comment by dschwen** [comment]
Argh, obviously for `std::complex` no greater/lesser operators are implemented.

**Comment by dschwen** [comment]
Ok, I can just add empty specializations to disable the members that require ordered comparisons. In th the future it might be nice to factor out a `RankTwoTensorTemplBase`, that has only operators and members that work with complex instantiation, and then derive `RankTwoTensorTempl` an add all the positive decomposition, givens rotation, eigenvalue stuff, etc. that only works with `Real` and `DualReal`. What do you think @lindsayad ?

**Comment by dschwen** [comment]
Also: I'm getting some crazy failures on "Boost enabled, pthreads only" - `Illegal instruction`. WTH? Any ideas @milljm ?

**Comment by milljm** [comment]
@dschwen I can't seem to get this to fail using moosebuild on Rod. I'm guessing something's fishy with mobuild machine it was running on...

**Comment by milljm** [comment]
Cool. Looks like I was right. Just one of those pesky older mobuild machines.

File: framework/include/utils/RankFourTensorImplementation.h [hunk 1]
```diff
@@ -0,0 +1,902 @@
...
+
+// C++ includes
+#include <iomanip>
+#include <ostream>
+
+namespace MathUtils
+{
+template <>
+void
+mooseSetToZero<RankFourTensorTempl<Real>>(RankFourTensorTempl<Real> & v)
```
Comment (lindsayad) [inline framework/include/utils/RankFourTensorImplementation.h]: As a full template specialization, won't this cause ODR failures if this header ever gets included in multiple translation units? Seems like maybe the best thing to do would be to move this definition into `RankFourTensor.C`? Same for the `DualReal` version as well as analogs in `RankThreeTensor` and `RankTwoTensor`.

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
> Ok, I can just add empty specializations to disable the members that require ordered comparisons. In th the future it might be nice to factor out a `RankTwoTensorTemplBase`, that has only operators and members that work with complex instantiation, and then derive `RankTwoTensorTempl` an add all the positive decomposition, givens rotation, eigenvalue stuff, etc. that only works with `Real` and `DualReal`. What do you think @lindsayad ?

I think that this sounds very reasonable.

File: framework/include/utils/RankFourTensorImplementation.h (hunk 1 above)
Comment (dschwen) [reply framework/include/utils/RankFourTensorImplementation.h]: But... we have `#pragma once` :-)

Comment (lindsayad) [reply framework/include/utils/RankFourTensorImplementation.h]: > multiple translation units?

is the key here. Unless you're campaigning for a `#pragma program_once`? 😄

Comment (dschwen) [reply framework/include/utils/RankFourTensorImplementation.h]: Oh, not in the implementation. Fixing it properly.

**Comment by dschwen** [comment]
@lindsayad the further I progress with factoring out a RankTwoTensorBase template the more of a mess I think it becomes. All tensor operators need to work on both the base and the derived RankTwoTensor templates specialized on Real and DualReal. The code gets hader and harder to undestand. I wonder if it wouldn't make more sense to just pull out the comparison dependent functionality into a utility namespace.

**Comment by lindsayad** [comment]
I trust you’ll do what is best :-)

**Comment by dschwen** [comment]
Ugh, is it not guaranteed that we have Boost type traits? Ok, well, I guess I can cook up my own "is comparable" trait...

**Comment by dschwen** [comment]
Ok, to sum up, I scrapped the base class idea because it led to a lot of code duplication (to make sure the operators do not down grade results to the base class, which, when captured with `auto` would lead to "missing" member functions).
I'm using "enable if" to some enable methods only for Real and DualReal. For other underlying types these methods will return a runtime error.

File: framework/include/utils/RankTwoTensor.h [hunk 2]
```diff
@@ -279,8 +280,12 @@ class RankTwoTensorTempl : public TensorValue<T>
   RankFourTensorTempl<T> mixedProductIlJk(const RankTwoTensorTempl<T> & a) const;
 
   /// return positive projection tensor of eigen-decomposition
-  RankFourTensorTempl<T> positiveProjectionEigenDecomposition(std::vector<T> & eigval,
-                                                              RankTwoTensorTempl<T> & eigvec) const;
+  template <typename T2 = T>
+  typename std::enable_if<MooseUtils::IsLikeReal<T2>::value, RankFourTensorTempl<T>>::type
```
Comment (lindsayad) [inline framework/include/utils/RankTwoTensor.h]: I **love** this SFINAE pattern implementation. I'm using it liberally now on https://github.com/libMesh/libmesh/pull/2416.

Is this ready for review?

Comment (dschwen) [reply framework/include/utils/RankTwoTensor.h]: Yes, I think so. By the way, we have a bunch of boostcopy::enable_if_c in moose. Should we just move to std::enable_if everywhere?

Comment (lindsayad) [reply framework/include/utils/RankTwoTensor.h]: We definitely could

**Comment by lindsayad** [comment]
@dschwen I'm a little concerned about the increase in time on the PETSc submodule module testing. For the past 20 or so events on the MOOSE CIVET next branch, there has been no instance where the entire target took longer than 2 hours to run, but here we are taking 2:40 for just the module testing (which admittedly is most of the target). It's possible that this is just a fluke, but it doesn't seem statistically likely. Do you have any reason to think this PR should slow anything down? I am more inclined to believe this is a fluke because one of the timeouts was out of navier stokes which does nothing with `Rank*Tensor`

File: framework/include/utils/RankFourTensorImplementation.h [hunk 3]
```diff
@@ -0,0 +1,894 @@
...
+// C++ includes
+#include <iomanip>
+#include <ostream>
+
+namespace MathUtils
+{
+template <>
+void mooseSetToZero<RankFourTensorTempl<Real>>(RankFourTensorTempl<Real> & v);
+template <>
+void mooseSetToZero<RankFourTensorTempl<DualReal>>(RankFourTensorTempl<DualReal> & v);
```
Comment (lindsayad) [inline framework/include/utils/RankFourTensorImplementation.h]: These are repeat declarations that are already in `RankFourTensor.h`. So you can remove them if you would like. It's not doing any harm here though, so I won't require it :smile: The same is true for the analogs in `RankThreeTensor` and `RankFourTensor`

**Review by lindsayad — COMMENTED** [review]
I'm ready to approve when we understand the increase in timing for the PETSc submodule module testing target

**Comment by dschwen** [comment]
> @dschwen I'm a little concerned about the increase in time on the PETSc submodule module testing. For the past 20 or so events on the MOOSE CIVET next branch, there has been no instance where the entire target took longer than 2 hours to run, but here we are taking 2:40 for just the module testing (which admittedly is most of the target).

I think this may be related to the power issues in EROB. Maybe testing was migrated to different boxes (@permcody  or @milljm)? None of the changes should affect the code that is generated for the Real/DualReal instantiations in the end (I think).

**Comment by lindsayad** [comment]
Oh wow beautiful passing all tests

**Review by lindsayad — APPROVED** [review]

**Pull request merged**